third_party/stdlib/test/test_contextlib.py

```diff
@@ -0,0 +1,353 @@
+"""Unit tests for contextlib.py, and other context managers."""
+
+import sys
+import tempfile
+import unittest
+# from contextlib import *  # Tests __all__
+import contextlib
+contextmanager = contextlib.contextmanager
+nested = contextlib.nested
+closing = contextlib.closing
+from test import test_support
+try:
+    import threading
+except ImportError:
+    threading = None
+
+
+class ContextManagerTestCase(unittest.TestCase):
+
+    def test_contextmanager_plain(self):
+        state = []
+        @contextmanager
+        def woohoo():
+            state.append(1)
+            yield 42
+            state.append(999)
+        with woohoo() as x:
+            self.assertEqual(state, [1])
+            self.assertEqual(x, 42)
+            state.append(x)
+        self.assertEqual(state, [1, 42, 999])
+
+    @unittest.skip('grumpy')
```

Comment:
Note that Grumpy has a built in called \_\_frame\_\_() that gets the top level fame and you can use f_back to get subsequent ones.